trieste/acquisition/rule.py

```diff
@@ -1103,6 +1115,361 @@ def state_func(
         return state_func
 
 
+class UpdateableSearchSpace(SearchSpace):
+    """A search space that can be updated."""
+
+    @abstractmethod
+    def reinitialize(
+        self,
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Reinitialize the search space using the given models and datasets."""
+        ...
+
+    @abstractmethod
+    def update(
+        self,
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Update the search space using the given models and datasets."""
+        ...
+
+    def get_single_model_and_dataset(
+        self,
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> Tuple[Optional[ProbabilisticModelType], Optional[Dataset]]:
+        """Get a single model and dataset from the given arguments."""
+        if models is None:
+            model = None
+        elif OBJECTIVE in models.keys():
+            model = models[OBJECTIVE]
+        else:
+            raise ValueError(f"""if models is provided, it must contain the key {OBJECTIVE}""")
+
+        if datasets is None:
+            dataset = None
+        elif OBJECTIVE in datasets.keys():
+            dataset = datasets[OBJECTIVE]
+        else:
+            raise ValueError(f"""if datasets is provided, it must contain the key {OBJECTIVE}""")
+
+        return model, dataset
+
+
+UpdateableSearchSpaceType = TypeVar("UpdateableSearchSpaceType", bound=UpdateableSearchSpace)
+""" A type variable bound to :class:`UpdateableSearchSpace`. """
+
+
+class MultiTrustRegion(
+    AcquisitionRule[
+        types.State[Optional["MultiTrustRegion.State"], TensorType],
+        SearchSpace,
+        ProbabilisticModelType,
+    ],
+    Generic[ProbabilisticModelType, UpdateableSearchSpaceType],
+):
+    """Abstract class for multi trust region acquisition rules. These are batch algorithms where
+    each query point is optimized in parallel, with its own separate trust region.
+    """
+
+    @dataclass(frozen=True)
+    class State:
+        """The acquisition state for the :class:`MultiTrustRegion` acquisition rule."""
+
+        acquisition_space: TaggedMultiSearchSpace
+        """ The search space. """
+
+        def __deepcopy__(self, memo: dict[int, object]) -> MultiTrustRegion.State:
+            acquisition_space_copy = copy.deepcopy(self.acquisition_space, memo)
+            return MultiTrustRegion.State(acquisition_space_copy)
+
+    def __init__(
+        self: "MultiTrustRegion[ProbabilisticModelType, UpdateableSearchSpaceType]",
+        subspace_type: Type[UpdateableSearchSpaceType],
```

Comment:
(sorry if this comment comes twice; i thought i submitted it but it seems to have disappeared!)

This seems like a code smell! Is a runtime subspace_type argument only required for the default implementation of create_subspace? If so, why does the ABC even need a default implementation?

Reply:
It is also used in the `MultiTrustRegionBox` class (recent change), not just the default class. The default implementation is just for simple cases where the update-able search space doesn't take any arguments, and maybe save some typing. However, I agree that is unlikely to be very useful.

The sole reason `subspace_type` exists is for use as a factory, to create the correct update-able search space; potentially allowing the users to only subclass the update-able search space and re-use an existing rule.

However, I am open to suggestions for a better mechanism for this.

Reply:
Another way of asking the same question: why can't we just

1. remove subspace_type from the initialisers
2. keep create_subpace abstract (no default implementation)
3. make MultiTrustRegionBox.create_subspace create and return a TrustRegionBox